File: Helper_Scripts/bmp2rgb565.py
# Converts a bmp file to a RGB565 array

import argparse
from PIL import Image

def rgb_to_rgb565(r, g, b):
    return ((r & 0xF8) << 8) | ((g & 0xFC) << 3) | (b >> 3)

def convert_bmp_to_rgb565_array(file_path):
    # Open image
    image = Image.open(file_path)
    
    # Retrieve pixel data
    pixels = list(image.convert("RGB").getdata())

    # Initialize RGB565 array
    rgb565_array = []

    # Iterate through each pixel and convert to RGB565
    for r, g, b in pixels:
        rgb565_value = rgb_to_rgb565(r, g, b)
        rgb565_array.append(str(rgb565_value))

    # Join the array elements into a string with square brackets
    return "[" + ", ".join(rgb565_array) + "]"

# Create argument parser
parser = argparse.ArgumentParser(description="Convert a BMP file to an RGB565 array string.")
parser.add_argument("file_path", help="Path to the BMP file to be converted.")

# Parse arguments
args = parser.parse_args()

# Convert the BMP file to an RGB565 array string
rgb565_array_string = convert_bmp_to_rgb565_array(args.file_path)

# Print the converted string
print(rgb565_array_string)

File: Helper_Scripts/bmp2rgb565.go
package main

import (
	"fmt"
	"image"
	_ "image/png" // Importing for PNG decoding
	"os"
)

func rgbToRGB565(r, g, b uint32) uint16 {
	return uint16(((r & 0xF8) << 8) | ((g & 0xFC) << 3) | (b >> 3))
}

func convertBmpToRGB565Array(filePath string) ([]uint16, error) {
	// Open the BMP file
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	// Decode the BMP image
	imageData, _, err := image.Decode(file)
	if err != nil {
		return nil, err
	}

	// Get the dimensions of the image
	bounds := imageData.Bounds()
	width, height := bounds.Max.X, bounds.Max.Y

	// Initialize the RGB565 array
	rgb565Array := make([]uint16, 0, width*height)

	// Iterate through each pixel and convert to RGB565
	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			r, g, b, _ := imageData.At(x, y).RGBA()
			rgb565Value := rgbToRGB565(r>>8, g>>8, b>>8)
			rgb565Array = append(rgb565Array, rgb565Value)
		}
	}

	return rgb565Array, nil
}

func main() {
	// Retrieve the file path from command-line arguments
	if len(os.Args) < 2 {
		fmt.Println("File path argument is required.")
		return
	}
	filePath := os.Args[1]

	// Convert the BMP file to an RGB565 array
	rgb565Array, err := convertBmpToRGB565Array(filePath)
	if err != nil {
		fmt.Printf("Error: %s\n", err)
		return
	}

	// Print the converted array
	fmt.Printf("%v\n", rgb565Array)
}
